torch/utils/data/dataloader.py

```diff
@@ -1034,8 +1044,24 @@ def __init__(self, loader):
             #     before it starts, and __del__ tries to join but will get:
             #     AssertionError: can only join a started process.
             w.start()
-            self._index_queues.append(index_queue)
-            self._workers.append(w)
+            return index_queue, w
+        if should_thread:
+            import concurrent.futures
+            with concurrent.futures.ThreadPoolExecutor(max_workers=self._num_workers) as executor:
```

Comment:
Do this work with `spawn`?
Could you help us try to use `multiprocessing_ctx="spawn"` in `DataLoader`?

Reply:
I've tested it with all start modes. In fact, as far as I can tell, the performance issue is only present with "spawn" and "forkserver" to begin with. 

By the way, this is something that puzzles me. The data loader uses "fork" by default. With "fork", there's no performance issue, but it is not a legal start mode in a GPU-enabled PyTorch, because forking after GPU initialization results in undefined behavior both in CUDA and ROCm. The documentation says so explicitly: https://github.com/pytorch/pytorch/blob/master/docs/source/notes/multiprocessing.rst I ran into the issue in the first place when I tried to switch to "spawn", because I was getting GPU-related errors with "fork".

Reply:
I believe some non-picklable `Dataset` can still rely on `fork` to do the multiprocessing without actual pickle the `Dataset`. This implementation would violate users' implementation.

If you test an example of `Dataset` with opened file handles, it should work with `fork`. But, after this change, as we always do `pickle.dumps`, a PickleError will be raised.

> The data loader uses "fork" by default.

This is incorrect. The starting method is platform specific. Like on windows, `spawn` is the default method and `fork` is not even available.

Reply:
If the dataset can't be pickled, the exception will be caught (lines 1020-1023) and it will revert to original behavior.